src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        if (Array.isArray(action.keysToRemove)) {
+          currentState = { ...currentState };
+        }
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
```

Comment:
The code assumes `action.keysToRemove` is always defined and an array. Although there is a check for it being an array, you should also ensure it exists before iterating over it to prevent potential runtime errors. Consider adding a condition to check if `action.keysToRemove` is defined and is an array before the loop.